concordium-consensus/src/Concordium/GlobalState/Persistent/TreeState.hs

```diff
@@ -880,12 +898,26 @@ instance
                 oldCredential <- case wmdData of
                     CredentialDeployment{} -> memberTransactionTable wmdHash
                     _ -> return False
-                let ~(added, newTT) = addTransaction bi 0 verRes tt
-                if not oldCredential && added
+                -- We need to check here that the nonce is still ok with respect to the last finalized block,
+                -- because it could be that a block was finalized thus the next account nonce being incremented
+                -- after this transaction was received and pre-verified.
+                isNonceOk <- case wmdData of
+                    NormalTransaction tr -> do
+                        lfbState <- use (skovPersistentData . lastFinalized . to _bpState)
+                        mAcc <- getAccount lfbState $ transactionSender tr
+                        nonce <- maybe (pure minNonce) getAccountNonce (snd <$> mAcc)
+                        return $! nonce <= transactionNonce tr
+                    -- the sequence number will be checked by @Impl.addTransaction@.
+                    _ -> return False
+                if isNonceOk
```

Comment:
I would suggest combining the `oldCredential` and `isNonceOk` checks into one check, since they are both casing on the `wmdData`. Actually, it seems very suspicious that we would add a credential deployment to the transaction table where it's already in the on-disk transaction table!!

Reply:
good catch.